pep-0681.rst

```diff
@@ -0,0 +1,709 @@
+  unspecified, ``init`` defaults to True. Field descriptor functions
+  can use overloads that implicitly specify the value of ``init``
+  using a literal bool value type (Literal[False] or Literal[True]).
+* ``default`` is an optional parameter that provides the default value
+  for the field.
+* ``default_factory`` is an optional parameter that provides a runtime
+  callback that returns the default value for the field. If neither
+  ``default`` nor ``default_factory`` are specified, the field is
+  assumed to have no default value and must be provided a value when
+  the class is instantiated.
+* ``factory`` is an alias for ``default_factory``. Stdlib dataclasses
+  use the name ``default_factory``, but attrs uses the name ``factory``
+  in many scenarios, so this alias is necessary for supporting attrs.
+* ``alias`` is an optional str parameter that provides an alternative
+  name for the field. This alternative name is used in the synthesized
+  ``__init__`` method.
+
+It is an error to specify more than one of ``default``,
+``default_factory`` and ``factory``.
+
+This example demonstrates::
+
+  # Library code (within type stub or inline)
+  # In this library, passing a resolver means that init must be False,
+  # and the overload with Literal[False] enforces that.
+  @overload
+  def model_field(
+          *,
+          default: Optional[Any] = ...,
+          resolver: Callable[[], Any],
+          init: Literal[False] = False,
+      ) -> Any: ...
+  
+  @overload
+  def model_field(
+          *,
+          default: Optional[Any] = ...,
+          resolver: None = None,
+          init: bool = True,
+      ) -> Any: ...
+  
+  @typing.dataclass_transform(
+      kw_only_default=True,
+      field_descriptors=(model_field, ))
+  def create_model(
+      *,
+      init: bool = True
+  ) -> Callable[[Type[_T]], Type[_T]]: ...
+  
+  
+  # Code that imports this library:
+  @create_model(init=False)
+  class CustomerModel:
+      id: int = model_field(resolver=lambda : 0)
+      name: str
+
+
+Runtime behavior
+----------------
+
+At runtime, the ``dataclass_transform`` decorator's only effect is to
+set a string attribute named ``__dataclass_transform__`` on the
+decorated function or class to support introspection. The value of the
+attribute should be a dict mapping the names of the
+``dataclass_transform`` parameters to their values.
+
+For example::
+
+  {
+    "eq_default": True,
+    "order_default": False,
+    "kw_only_default": False,
+    "field_descriptors": ()
+  }
+
+
+Dataclass semantics
+-------------------
+
+The following dataclass semantics are implied when
+``dataclass_transform`` is specified.
+
+1. Frozen classes cannot inherit from non-frozen classes. A class that
+   has been decorated with ``dataclass_transform`` will not be
+   considered non-frozen. Similarly, a class that directly specifies a
+   metaclass that has been decorated with ``dataclass_transform`` will
+   not be considered non-frozen.
```

Comment:
I'm working on improving the language here.

@msfterictraut, is there a PEP/document/etc which specifies that "frozen classes cannot inherit from non-frozen classes". I don't see that in [PEP 557](https://www.python.org/dev/peps/pep-0557/) or the [dataclass docs](https://docs.python.org/3/library/dataclasses.html)

Reply:
@CAM-Gerlach is the new language acceptable?

I waiting to resolve this until Eric Traut replies. Might be useful to have a link to whatever documentation there is on this constraint.

Reply:
I don't think this behavior is documented in PEP 557, but it is consistent with the runtime implementation. I didn't know about it until someone filed a bug against pyright.

Here's a code snippet from the pyright unit tests. If you run this in the Python interpreter, you'll see that it generates a runtime exception: "TypeError: cannot inherit frozen dataclass from a non-frozen one".

```python
from dataclasses import dataclass

@dataclass(frozen=False)
class DC1:
    val1: int = 6

# This should generate an error because a frozen dataclass
# cannot inherit from a non-frozen dataclass.
@dataclass(frozen=True)
class DC3(DC1):
    val3: int = 4
```

Reply:
Sorry, missed seeing this (I noticed the other open comment, but it wasn't that critical, just a passing question). Its a lot clearer now, thanks.

Reply:
I tracked this down to https://bugs.python.org/issue32960. Would it be worthwhile to include a link to this in the PEP?